src/constraint/4C_constraint_springdashpot.cpp

```diff
@@ -10,21 +10,228 @@
+            const Core::Elements::JacobianMapping<celltype, dim>& jacobian_mapping,
+            double integration_factor, int gp)
+        {
+          Core::LinAlg::Matrix<dim, 1> displacement_at_gp;
+          displacement_at_gp.multiply_tn(displacements, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> offset_at_gp;
+          offset_at_gp.multiply_tn(displacement_offset, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> velocity_at_gp;
+          velocity_at_gp.multiply_tn(velocity, shape_functions.values);
+
+          Core::LinAlg::Matrix<dim, 1> normal_at_gp = get_normal<celltype>(jacobian_mapping);
+
+          double displacement_ref_normal = displacement_at_gp.dot(normal_at_gp);
+          double offset_ref_normal = offset_at_gp.dot(normal_at_gp);
+          double velocity_ref_normal = velocity_at_gp.dot(normal_at_gp);
+
+          const auto& [force, force_deriv_disp, force_deriv_vel] = spring_dashpot_evaluator(
+              displacement_ref_normal - offset_ref_normal - constant_offset, velocity_ref_normal);
+
+          // evaluate residual force
+          if (residual_vector)
+          {
+            for (unsigned axis_direction = 0; axis_direction < dim; ++axis_direction)
+            {
+              for (int node = 0; node < Core::FE::num_nodes<celltype>; ++node)
+              {
+                residual_vector.value()(node * dim + axis_direction) +=
+                    shape_functions.values(node) * force * normal_at_gp(axis_direction) *
+                    integration_factor;
+              }
+            }
+          }
+
+          // evaluate stiffness matrix
+          if (stiffness_matrix)
+          {
+            for (unsigned axis_direction1 = 0; axis_direction1 < dim; ++axis_direction1)
+            {
+              for (unsigned axis_direction2 = 0; axis_direction2 < dim; ++axis_direction2)
+              {
+                for (int i = 0; i < Core::FE::num_nodes<celltype>; ++i)
+                {
+                  for (int j = 0; j < Core::FE::num_nodes<celltype>; ++j)
+                  {
+                    stiffness_matrix.value()(
+                        i * dim + axis_direction1, j * dim + axis_direction2) +=
+                        shape_functions.values(i) * shape_functions.values(j) *
+                        (force_deriv_disp + force_deriv_vel * time_factor) *
+                        normal_at_gp(axis_direction1) * normal_at_gp(axis_direction2) *
+                        integration_factor;
+                  }
+                }
+              }
+            }
+          }
+        });
+  }
+
+  template <Core::FE::CellType celltype, int num_dof_per_node>
+  Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, num_dof_per_node> get_dof_matrix(
+      const std::vector<double>& values)
+  {
+    Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, num_dof_per_node> dof_matrix;
```

Comment:
This function should ideally perform a `memcpy` from values into dof_matrix.A(), right? This is what https://github.com/4C-multiphysics/4C/blob/main/src/core/linalg/src/dense/4C_linalg_fixedsizematrix.hpp#L1984 does. You might want to use a view here.